flink-table/flink-table-planner/src/main/java/org/apache/flink/table/plan/rules/logical/FlinkAggregateExtractProjectRule.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.calcite.rel.logical.LogicalAggregate;
+import org.apache.calcite.rel.rules.AggregateExtractProjectRule;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.mapping.Mapping;
+import org.apache.calcite.util.mapping.MappingType;
+import org.apache.calcite.util.mapping.Mappings;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Rule to extract a {@link org.apache.calcite.rel.core.Project}
+ * from a {@link LogicalAggregate} or a {@link LogicalWindowAggregate}
+ * and push it down towards the input.
+ *
+ * <p>Note: Most of the logic in this rule is same with {@link AggregateExtractProjectRule}. The
+ * difference is this rule has also taken the {@link LogicalWindowAggregate} into consideration.
+ */
+public class FlinkAggregateExtractProjectRule extends AggregateExtractProjectRule {
+
+	public FlinkAggregateExtractProjectRule(RelOptRuleOperand operand, RelBuilderFactory builderFactory) {
+		super(operand, builderFactory);
+	}
+
+	private int getWindowTimeFieldIndex(LogicalWindowAggregate aggregate, RelNode input) {
+		LogicalWindow logicalWindow = aggregate.getWindow();
+		ResolvedFieldReference timeAttribute = (ResolvedFieldReference) logicalWindow.timeAttribute();
+		return input.getRowType().getFieldNames().indexOf(timeAttribute.name());
+	}
+
+	@Override
+	public boolean matches(RelOptRuleCall call) {
+		final Aggregate aggregate = call.rel(0);
+		return aggregate instanceof LogicalWindowAggregate || aggregate instanceof LogicalAggregate;
+	}
+
+	@Override
+	public void onMatch(RelOptRuleCall call) {
+		final Aggregate aggregate = call.rel(0);
+		final RelNode input = call.rel(1);
+		// Compute which input fields are used.
+		// 1. group fields are always used
+		final ImmutableBitSet.Builder inputFieldsUsed =
+			aggregate.getGroupSet().rebuild();
+		// 2. agg functions
+		for (AggregateCall aggCall : aggregate.getAggCallList()) {
+			for (int i : aggCall.getArgList()) {
+				inputFieldsUsed.set(i);
+			}
+			if (aggCall.filterArg >= 0) {
+				inputFieldsUsed.set(aggCall.filterArg);
+			}
+		}
+		// 3. window time field if the aggregate is a group window aggregate.
+		if (aggregate instanceof LogicalWindowAggregate) {
+			inputFieldsUsed.set(getWindowTimeFieldIndex((LogicalWindowAggregate) aggregate, input));
+		}
+
+		final RelBuilder relBuilder = call.builder().push(input);
+		final List<RexNode> projects = new ArrayList<>();
+		final Mapping mapping =
+			Mappings.create(MappingType.INVERSE_SURJECTION,
+				aggregate.getInput().getRowType().getFieldCount(),
+				inputFieldsUsed.cardinality());
+		int j = 0;
+		for (int i : inputFieldsUsed.build()) {
+			projects.add(relBuilder.field(i));
+			mapping.set(i, j++);
+		}
+
+		if (input instanceof Project && input.getRowType().getFieldCount() == projects.size()) {
```

Comment:
Is it correct? What if the types mismatch? I think just comparing input size is not enough. Also `relBuilder.project(projects);` handles trivial projections internally.

Reply:
It seems there is no need to add the trivial project here. The input is also a project and the trivial project hasn't performed any column pruning. Or have I missed something?

Reply:
Yes, you are right it should work. However this is handled already in `relBuilder.project()` without the `force` flag.